core/src/main/java/jenkins/console/ConsoleURLProvider.java

```diff
@@ -0,0 +1,90 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.console;
+
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+import hudson.Functions;
+import hudson.model.Run;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.Stapler;
+
+/**
+ * Extension point that allows implementations to redirect build console links to a specified URL.
+ * <p>In order to produce links to console URLs in Jelly templates, use {@link Functions#getConsoleUrl()}.
+ * @see Functions#getConsoleUrl
+ * @since TODO
+ */
+public interface ConsoleUrlProvider extends ExtensionPoint {
+    @Restricted(NoExternalUse.class)
+    Logger LOGGER = Logger.getLogger(ConsoleUrlProvider.class.getName());
+
+    /**
+     * Get a URL relative to the context path of Jenkins which should be used to link to the console for the specified build.
+     * <p>Should only be used in the context of serving an HTTP request.
+     * @param run the build
+     * @return the URL for the console for the specified build, relative to the context of Jenkins, or {@code null}
+     * if this implementation does not want to server a special console view for this build.
+     */
+    @CheckForNull String getConsoleUrl(Run<?, ?> run);
+
+    /**
+     * Get a URL relative to the web server root which should be used to link to the console for the specified build.
+     * <p>Should only be used in the context of serving an HTTP request.
+     * <p>Use {@link Functions#getConsoleUrl} to obtain this link in a Jelly template.
+     * @param run the build
+     * @return the URL for the console for the specified build, relative to the web server root
+     */
+    static @NonNull String getRedirectUrl(Run<?, ?> run) {
+        String url = null;
+        for (ConsoleUrlProvider provider : ExtensionList.lookup(ConsoleUrlProvider.class)) {
```

Comment:
@daniel-beck raised a few important points in https://github.com/jenkinsci/jenkins/pull/8321#pullrequestreview-1563282385 and I am not entirely sure how we should address them.

With my latest changes to the API, we can create implementations that only care about Pipelines, or builds with odd numbers, etc. However, if both Blue Ocean and Pipeline Graph View wanted to implement `ConsoleUrlProvider` for all Pipeline builds, and you have both installed on a single controller, they would compete for precedence based on `ordinal`. How do we want to address this? I can see a few approaches:
* Do nothing, let admins use https://plugins.jenkins.io/extension-filter/ as needed. Maybe this is good enough if we only expect a few overlapping implementations of this extension point, especially if we think the implementations are likely to come from competing plugins where users are not likely to them at the same time.
* Provide a `GlobalConfiguration` UI in core to allow admins to select which `ConsoleUrlProvider` extensions should be enabled on a given controller, or allow them to be sorted based on the admin's preference. Same end result as above, though arguably more user-friendly, but also it has additional implementation complexity.
* Provide a user-based configuration in core to let users select which `ConsoleUrlProvider` extensions they want to allow, or allow them to be sorted based on the user's preference. Adds even more implementation complexity than the above options, but now if there are two plugins installed with competing `ConsoleUrlProvider` implementations, individual users get to pick which one they want to use.

For now at least I need to update https://github.com/jenkinsci/pipeline-graph-view-plugin/pull/288 and https://github.com/jenkinsci/pipeline-input-step-plugin/pull/145 based on the latest changes.

Reply:
I suspect you will want 2 and 3, admins to set a default and allow users to override it

Reply:
Missed this thread and posted https://github.com/jenkinsci/jenkins/pull/8321#issuecomment-1666464680. For now IMO option 2 should be enough (option 3 only if sufficient user demand exists).

Reply:
I feel you'll get users that are used to it but some users who would want the other implementation, also with the display url only having a per user setting it seems weird for this one not to?

Reply:
For a cloudbees plugin it may make more sense as a control of the whole instance.

In the case of pipeline graph view plugin someone may want to fallback to the default console view if it doesn't handle something for them or they have used it for 10+ years and do not want to change.

also PGV doesn't support freestyle builds at all and I haven't seen any handling for that in the downstream PR.

Reply:
> also PGV doesn't support freestyle builds at all and I haven't seen any handling for that in the downstream PR.

Yes sorry, all of the downstream PRs need to be updated. I will get to that later today.

Reply:
See [018789a](https://github.com/jenkinsci/jenkins/pull/8321/commits/018789a84712f3f112b69dae6d0739325a26adcb) and let me know what you think. We can dial it back as desired.

Reply:
this looks good to me